kyo-prelude/shared/src/main/scala/kyo2/Abort.scala

```diff
@@ -0,0 +1,101 @@
+package kyo2
+
+import kernel.<
+import kernel.Const
+import kernel.Effect
+import kernel.Frame
+import kyo.Tag
+import scala.annotation.targetName
+import scala.reflect.ClassTag
+import scala.util.Failure
+import scala.util.Success
+import scala.util.Try
+
+sealed trait Abort[-E] extends Effect[Const[E], Const[Unit]]
+
+object Abort:
+
+    private inline def erasedTag[E]: Tag[Abort[E]] = Tag[Abort[Any]].asInstanceOf[Tag[Abort[E]]]
+
+    inline def fail[E](inline value: E): Nothing < Abort[E] =
+        Effect.suspendMap[Any](erasedTag[E], value)(_ => ???)
+
+    inline def when[E](b: Boolean)(inline value: E)(
+        using inline tag: Tag[Abort[E]]
+    ): Unit < Abort[E] =
+        if b then fail(value)
+        else ()
+
+    class GetOps[E >: Nothing](dummy: Unit) extends AnyVal:
+        inline def apply[A](either: Either[E, A]): A < Abort[E] =
+            either match
+                case Right(value) => value
+                case Left(value)  => fail(value)
+
+        inline def apply[A](opt: Option[A]): A < Abort[None.type] =
+            opt match
+                case None    => fail(None)
+                case Some(v) => v
+
+        inline def apply[A](e: Try[A]): A < Abort[Throwable] =
+            e match
+                case Success(t) => t
+                case Failure(v) => fail(v)
+
+        inline def apply[A](r: Result[A]): A < Abort[Throwable] =
+            r.fold(fail)(identity)
+
+        @targetName("maybe")
+        inline def apply[A](m: Maybe[A]): A < Abort[Maybe.Empty] =
+            m.fold(fail(Maybe.Empty))(identity)
+    end GetOps
+
+    inline def get[E >: Nothing]: GetOps[E] = GetOps(())
+
+    class RunOps[E >: Nothing](dummy: Unit) extends AnyVal:
+        def apply[E0 <: E, A, B, ES, ER](v: A < (Abort[ES] & B))(
+            using
+            h: HasAbort[E0, ES] { type Remainder = ER },
+            ct: ClassTag[E],
+            frame: Frame
+        ): Either[E, A] < (ER & B) =
+            Effect.catching {
+                Effect.handle(erasedTag[E], v.map(Right(_): Either[E, A]))(
+                    accept = [C] =>
+                        (input, _) =>
+                            (input.asInstanceOf[Any]) match
+                                case input: E => true
+                                case _        => false,
+                    handle = [C] => (input, _) => Left(input)
+                ).asInstanceOf[Either[E, A] < ER & B]
+            } {
+                case fail: E => Left(fail)
```

Comment:
Should `catching` only be used if `E` is a subtype of `Throwable`? I wouldn't think an unbounded type parameter would work in this case.

Reply:
I think so since it can only recover thrown exceptions. The type is bound to `Throwable` here and in `Effect.catching`.

Reply:
I see, thanks for the clarification! It's safe to do it even without the bound since the tag will simply mismatch when catching exceptions. That said, maybe we should constrain `Abort` only to exception types. It'd allow us to inject the trace into the stack trace for example